### PR TITLE
[Merged by Bors] - feat(Algebra/Category/ModuleCat/Presheaf): presheaves of modules are generated by their sections

### PR DESCRIPTION
**Pull request by joelriou**
We provide a presentation of any presheaf of modules involving coproducts of free presheaves of modules generated by Yoneda presheaves of types.

---

- [x] depends on: #18159

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Algebra/Category/ModuleCat/Presheaf/Generator.lean [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+
+lemma freeYonedaEquiv_comp {M N : PresheafOfModules.{v} R} {X : C}
+    (m : ((free R).obj (yoneda.obj X) ⟶ M)) (φ : M ⟶ N) :
+    freeYonedaEquiv (m ≫ φ) = φ.app _ (freeYonedaEquiv m) := rfl
+
+variable (R) in
+/-- The set of `PresheafOfModules.{v} R` consisting of objects of the
+form `(free R).obj (yoneda.obj X)` for some `X`.  -/
+inductive FreeYoneda : Set (PresheafOfModules.{v} R)
+  | free_yoneda_mem (X : C) : FreeYoneda ((free R).obj (yoneda.obj X))
```
Comment (erdOne) [inline Mathlib/Algebra/Category/ModuleCat/Presheaf/Generator.lean:61]: This sees to be abusing the `Set = _ -> Prop` defeq. Either replace it with `{ (free R).obj (yoneda.obj X) | (X : C) }` (if it works; not tested) and renamed to `freeYoneda`, or define it as `_ -> Prop` and use `setOf` below.

Comment (joelriou) [reply Mathlib/Algebra/Category/ModuleCat/Presheaf/Generator.lean]: Thanks for the suggestion. I have just fixed this.

File: Mathlib/Algebra/Category/ModuleCat/Presheaf/Generator.lean [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+
+lemma isDetecting : IsDetecting (freeYoneda R) :=
+  (isSeparating R).isDetecting
+
+end freeYoneda
+
+instance wellPowered : WellPowered (PresheafOfModules.{u} R₀) :=
+  wellPowered_of_isDetecting (freeYoneda.isDetecting R₀)
+
+/-- The type of elements of a presheaf of modules. -/
```
Comment (kbuzzard) [inline Mathlib/Algebra/Category/ModuleCat/Presheaf/Generator.lean:85]:
```suggestion
/-- The type of elements of a presheaf of modules. A term of this type is a pair
`⟨X, a⟩` with `X : Cᵒᵖ` and `a : M.obj X`. -/
```

File: Mathlib/Algebra/Category/ModuleCat/Presheaf/Generator.lean [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+
+variable {C : Type u} [SmallCategory.{u} C] {R : Cᵒᵖ ⥤ RingCat.{u}} (M : PresheafOfModules.{u} R)
+
+/-- Given a presheaf of modules `M`, this is the coproduct of
+all free Yoneda presheaves `m.freeYoneda` for all `m : M.Elements`. -/
+noncomputable abbrev freeYonedaCoproduct : PresheafOfModules.{u} R :=
+  ∐ (Elements.freeYoneda (M := M))
+
+/-- Given an element `m : M.Elements` of a presheaf of modules `M`, this is the
+canonical inclusion `e.freeYoneda ⟶ M.freeYonedaCoproduct`. -/
```
Comment (kbuzzard) [inline Mathlib/Algebra/Category/ModuleCat/Presheaf/Generator.lean:126]:
```suggestion
canonical inclusion `m.freeYoneda ⟶ M.freeYonedaCoproduct`. -/
```

File: Mathlib/Algebra/Category/ModuleCat/Presheaf/Generator.lean [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+  Sigma.desc Elements.fromFreeYoneda
+
+/-- Given an element `m` of a presheaf of modules `M`, this is the associated
+canonical section of the presheaf `M.freeYonedaCoproduct` over the object `m.1`. -/
+noncomputable def freeYonedaCoproductMk (m : M.Elements) :
+    M.freeYonedaCoproduct.obj m.1 :=
+  (M.ιFreeYonedaCoproduct m).app _ (ModuleCat.freeMk (𝟙 _))
+
+@[reassoc (attr := simp)]
+lemma ι_fromFreeYonedaCoproduct (e : M.Elements) :
```
Comment (kbuzzard) [inline Mathlib/Algebra/Category/ModuleCat/Presheaf/Generator.lean:144]: Is there a reason you've switched notation from `m` to `e`? I don't have a preference but my instinct would be to be consistent.

File: Mathlib/Algebra/Category/ModuleCat/Presheaf/Generator.lean [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+lemma ι_fromFreeYonedaCoproduct_apply (e : M.Elements) (X : Cᵒᵖ) (x : e.freeYoneda.obj X) :
+    M.fromFreeYonedaCoproduct.app X ((M.ιFreeYonedaCoproduct e).app X x) =
+      e.fromFreeYoneda.app X x :=
+  congr_fun ((evaluation R X ⋙ forget _).congr_map (M.ι_fromFreeYonedaCoproduct e)) x
+
+@[simp]
+lemma fromFreeYonedaCoproduct_app_mk (e : M.Elements) :
+    M.fromFreeYonedaCoproduct.app _ (M.freeYonedaCoproductMk e) = e.2 := by
+  dsimp [freeYonedaCoproductMk]
+  erw [M.ι_fromFreeYonedaCoproduct_apply e]
```
Comment (kbuzzard) [inline Mathlib/Algebra/Category/ModuleCat/Presheaf/Generator.lean:157]: Do you know why `  rw [M.ι_fromFreeYonedaCoproduct_apply e e.fst (ModuleCat.freeMk (𝟙 (Opposite.unop e.fst)))]` doesn't work?

File: Mathlib/Algebra/Category/ModuleCat/Presheaf/Generator.lean [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+  simp [toFreeYonedaCoproduct]
+
+/-- (Colimit) cofork which gives a presentation of a presheaf of modules `M` using
+coproducts of free presheaves of modules on Yoneda presheaves. -/
+noncomputable abbrev freeYonedaCoproductsCokernelCofork :
+    CokernelCofork M.toFreeYonedaCoproduct :=
+  CokernelCofork.ofπ _ M.toFreeYonedaCoproduct_fromFreeYonedaCoproduct
+
+/-- If `M` is a presheaf of modules, the cokernel cofork
+`M.freeYonedaCoproductsCokernelCofork` is colimit, which means that
```
Comment (kbuzzard) [inline Mathlib/Algebra/Category/ModuleCat/Presheaf/Generator.lean:183]:
```suggestion
`M.freeYonedaCoproductsCokernelCofork` is a colimit, which means that
```

**Review by kbuzzard — APPROVED** [review]
Very elegant stuff!

File: Mathlib/Algebra/Category/ModuleCat/Presheaf/Generator.lean [hunk 7]
```diff
@@ -0,0 +1,199 @@
...
+
+lemma freeYonedaEquiv_comp {M N : PresheafOfModules.{v} R} {X : C}
+    (m : ((free R).obj (yoneda.obj X) ⟶ M)) (φ : M ⟶ N) :
+    freeYonedaEquiv (m ≫ φ) = φ.app _ (freeYonedaEquiv m) := rfl
+
+variable (R) in
+/-- The set of `PresheafOfModules.{v} R` consisting of objects of the
+form `(free R).obj (yoneda.obj X)` for some `X`.  -/
+def freeYoneda : Set (PresheafOfModules.{v} R) := Set.range (yoneda ⋙ free R).obj
+
```
[the remark this answers is not in the document]
Comment (joelriou) [reply Mathlib/Algebra/Category/ModuleCat/Presheaf/Generator.lean:61]: When applied on a `X`, the better (`dsimp`) form is `(free R).obj (yoneda.obj X)`.

File: Mathlib/Algebra/Category/ModuleCat/Presheaf/Generator.lean (hunk 4 above)
Comment (joelriou) [reply Mathlib/Algebra/Category/ModuleCat/Presheaf/Generator.lean]: Thanks. I have fixed these inconsistencies.

File: Mathlib/Algebra/Category/ModuleCat/Presheaf/Generator.lean [hunk 8]
```diff
@@ -0,0 +1,206 @@
...
+-/
+
+universe v v₁ u u₁
+
+open CategoryTheory Limits
+
+namespace PresheafOfModules
+
+variable {C : Type u} [Category.{v} C] {R : Cᵒᵖ ⥤ RingCat.{v}}
+  {C₀ : Type u} [SmallCategory C₀] {R₀ : C₀ᵒᵖ ⥤ RingCat.{u}}
```
[the remark this answers is not in the document]
Comment (joelriou) [reply Mathlib/Algebra/Category/ModuleCat/Presheaf/Generator.lean]: Thanks for the suggestion. I have removed the unnecessary global variables.

File: Mathlib/Algebra/Category/ModuleCat/Presheaf/Generator.lean [hunk 9]
```diff
@@ -0,0 +1,206 @@
...
+import Mathlib.Algebra.Homology.ShortComplex.Exact
+import Mathlib.CategoryTheory.Elements
+import Mathlib.CategoryTheory.Generator
+
+/-!
+# Generators for the category of presheaves of modules
+
+In this file, given a presheaf of rings `R` on a category `C`,
+we construct a set `FreeYoneda R` of presheaves of modules that
+is separating and detecting: this set consists of presheaves of
```
[the remark this answers is not in the document]
Comment (joelriou) [reply Mathlib/Algebra/Category/ModuleCat/Presheaf/Generator.lean]: Thanks for the reviews! I have rewritten the module docstring.

File: Mathlib/Algebra/Category/ModuleCat/Presheaf/Generator.lean (hunk 5 above)
Comment (joelriou) [reply Mathlib/Algebra/Category/ModuleCat/Presheaf/Generator.lean]: This is basically the same reason why ` ι_fromFreeYonedaCoproduct_apply` cannot be a simp lemma, as, according to the linter, it is not in simpNF. If we apply `dsimp` to the statement, for the "human" eye, nothing changes, but there are hidden changes behind a `DFunLike.coe`: the definition of some types can be unfolded by dsimp. See also https://leanprover.zulipchat.com/#narrow/channel/287929-mathlib4/topic/attribute.20simp_NF_my_lemma.3F/near/443101462

File: Mathlib/Algebra/Category/ModuleCat/Presheaf/Generator.lean [hunk 10]
```diff
@@ -0,0 +1,213 @@
...
+/-- Given a presheaf of modules `M`, this is a constructor for the type `M.Elements`. -/
+abbrev elementsMk {C : Type u₁} [Category.{v₁} C] {R : Cᵒᵖ ⥤ RingCat.{u}}
+    (M : PresheafOfModules.{v} R) (X : Cᵒᵖ) (x : M.obj X) : M.Elements :=
+  Functor.elementsMk _ X x
+
+namespace Elements
+
+variable {C : Type u} [Category.{v} C] {R : Cᵒᵖ ⥤ RingCat.{v}} {M : PresheafOfModules.{v} R}
+
+/-- Given an element `m : M.Elements` of a presheaf of modules `M`, this is
```
Comment (TwoFX) [inline Mathlib/Algebra/Category/ModuleCat/Presheaf/Generator.lean:112]:
```suggestion
/-- Given an element `m : M.Elements` of a presheaf of modules `M`, this is the
```

**Review by TwoFX — APPROVED** [review]
Thanks!
bors d+

**Comment by joelriou** [comment]
Thanks!

bors merge

**Pull request closed without merging**